### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by dimon222**
To address #33, #19 and remove now unnecessary kerberos variables. Any kind of auth (HTTPKerberosAuth, HTTPSPNEGOAuth, basic auth, etc) can be directly passed to auth in constructor of necessary app context. Another good thing is now we get session instead of atomic calls. That would eliminate unnecessary additional handshakes for kerberos.

In this sense example with typical HTTPKerberosAuth:
```
from yarn_api_client.history_server import HistoryServer
from requests_kerberos import HTTPKerberosAuth
history_server = HistoryServer('https://127.0.0.2:5678', auth=HTTPKerberosAuth())
```
Same/similar logic for any requests supported module for auth. GSSAPI is one of the most flexible.

### CODE REVIEW

File: itests/integration_test_resource_manager.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def setUpClass(self):
 
             if yarn_endpoint_uri.hostname and yarn_endpoint_uri.port:
                 self.configured = True
-                self.resourceManager = ResourceManager(yarn_endpoint_uri.hostname, yarn_endpoint_uri.port)
+                self.resourceManager = ResourceManager(yarn_endpoint_uri.hostname + ":" + str(yarn_endpoint_uri.port))
```
Comment (lresende) [inline itests/integration_test_resource_manager.py]: Could you please refactor this to be resource_manager

File: yarn_api_client/application_master.py [hunk 2]
```diff
@@ -15,17 +15,19 @@ class ApplicationMaster(BaseYarnAPI):
...
     :param int timeout: API connection timeout in seconds
-    :param boolean kerberos_enabled: Flag identifying is Kerberos Security has been enabled for YARN
+    :param AuthBase auth: Auth to use for requests
+    :param boolean verify: Either a boolean, in which case it controls whether
+    we verify the server's TLS certificate, or a string, in which case it must
+    be a path to a CA bundle to use. Defaults to ``True``
     """
-    def __init__(self, address=None, port=8088, timeout=30, kerberos_enabled=False):
-        if address is None:
+    def __init__(self, url=None, timeout=30, auth=None, verify=True):
```
Comment (lresende) [inline yarn_api_client/application_master.py]: could you please rename url to service_endpoint

File: yarn_api_client/base.py [hunk 3]
```diff
@@ -12,37 +12,57 @@ def __init__(self, response):
         self.data = response.json()
 
 
+class Uri(object):
```
Comment (lresende) [inline yarn_api_client/base.py]: Please use the existing Uri functionality `from .uri import Uri` which will allow you to use the code snip described in #33

File: setup.py [hunk 4]
```diff
@@ -30,10 +30,6 @@ def find_version(*file_paths):
         'requests>=2.7,<3.0',
     ],
 
-    extras_require = {
```
Comment (lresende) [inline setup.py]: this would rely on client application defining all dependencies, is that the desired behavior? is this mostly because we are relying on the client application to provide the `auth=HTTPKerberosAuth()`?

**Comment by lresende** [comment]
@kevin-bates Any other thoughts here?

File: setup.py (hunk 4 above)
Comment (dimon222) [reply setup.py]: Correct, basically whole kerberos change is just 2 lines in codebase. We can completely eliminate it by asking user to install it. Another possible aapproach is but keeping extra package, but have no code that actually uses imports from it, since it user actually injects auth

File: yarn_api_client/base.py (hunk 3 above)
Comment (dimon222) [reply yarn_api_client/base.py]: The problem with it is that user will be more limited with types of format for URL. I tried multiple ways, but method `.getUrl()` doesn't prepend defaunt protocol and port

Comment (kevin-bates) [reply yarn_api_client/base.py]: I agree that we should not perform our own parsing of `service_endpoint`.  (Btw, @lresende's code snippet for his `Uri` class is just a wrapper of `urlparse` that "converts" `path` to `api_endpoint`.)

```python
service_uri = Uri(service_endpoint)
self.scheme = service_uri.scheme or 'https'
self.hostname = service_uri.hostname
self.port = service_uri.port or '8088'
self.api_endpoint = service_uri.path or '/ws/v1/cluster'
```
@dimon222 - how does this not address defaulting of the protocol (scheme) or port?  Can you please clarify?

Btw, I think we'd want to the default scheme to be `http` since the port is not a well-known ssl port.

File: yarn_api_client/hadoop_conf.py [hunk 5]
```diff
@@ -62,34 +57,20 @@ def get_resource_manager_host_port():
...
-    if value is not None:
-        host, _, port = value.partition(':')
-        return host, port
-    else:
-        return None
+    return parse(config_path, prop_name)
 
 
-def get_webproxy_host_port():
+def get_webproxy_host_port(timeout=30, auth=None, verify=True):
```
[the remark this answers is not in the document]
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: I was thinking about switching test for RM to requests and utilize Auth/verify. Mainly because I'm not about behaviour of httplib when user is using some broken self-signed certificate. Same thing - will RM check work in kerberized environment if Auth is not passed?

[the remark this answers is not in the document]
Comment (kevin-bates) [reply yarn_api_client/hadoop_conf.py]: I don't know web protocols well enough to answer that question, but it seems like we shouldn't add these parameters until they're needed.  It looks like `check_is_active_rm()` is merely a ping against the `/cluster` endpoint, so I suspect the current code is sufficient.  If this is changed to use `requests`, then it does seem like `auth` and `verify` should probably be provided - but, until then, I would vote to remove unused parameters.

File: yarn_api_client/base.py (hunk 3 above)
Comment (dimon222) [reply yarn_api_client/base.py]: `Requests` requires full address with schema specified, not atomic for each component. If we decouple it manually, we have to construct URL ourselves back, which I'm not sure about

Comment (kevin-bates) [reply yarn_api_client/base.py]: Couldn't we use [`urlunparse`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlunparse) to reconstruct the fully qualified service endpoint?

```python
urlunparse((self.scheme, self.hostname + ':' + str(self.port), self.path, None, None, None))
```

File: itests/integration_test_resource_manager.py (hunk 1 above)
Comment (dimon222) [reply itests/integration_test_resource_manager.py]: Addressed. Let me know what u think.

File: yarn_api_client/hadoop_conf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: Sure, no problem. Adjusted.

File: yarn_api_client/hadoop_conf.py [hunk 6]
```diff
@@ -41,19 +37,18 @@ def check_is_active_rm(rm_web_host, rm_web_port):
     else:
         if response.getheader('Refresh', None) is not None:
             return False
-        return True
+    return True
 
 
-def get_resource_manager_host_port():
+def get_resource_manager_host_port(timeout=30, auth=None, verify=True):
```
[the remark this answers is not in the document]
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: Adjusted. Same as another comment.

File: yarn_api_client/application_master.py (hunk 2 above)
Comment (dimon222) [reply yarn_api_client/application_master.py]: Addressed

File: yarn_api_client/base.py (hunk 3 above)
Comment (dimon222) [reply yarn_api_client/base.py]: Added some logic with urlparse/urlunparse, let me know what u think.

**Comment by dimon222** [comment]
@lresende @kevin-bates I've addressed mentioned things + updated definition of RM to use list of any amount of endpoints for "failover out of the box".

File: yarn_api_client/base.py [hunk 7]
```diff
@@ -6,43 +6,62 @@
...
+
+            if self.service_uri.port:
+                self.service_uri_postfix = ':' + str(self.service_uri.port)
+            else:
+                self.service_uri_postfix = ''
+        else:
+            self.service_uri = None
+            self.service_uri_scheme = None
+            self.service_uri_hostname = None
+            self.service_uri_postfix = None
```
Comment (kevin-bates) [inline yarn_api_client/base.py]: Can we please follow @lresende's lead and create a local `Uri` class that encapsulates this?  I don't see the need for all the extra member variables.  This class would also be used by the test code and eliminates the `import` hassle.  The class can be defined in `base.py` or a new file.

The constructor should call `urlparse` and initialize default values - as Luciano's previous snippet laid out.

Please keep `port` separated from the `:` and add that in when constructing the url.  The class can have a `to_url(api_path)` method that constructs a url.

Why is the `path` used as the `service_uri_hostname` if no `hostname` is parsed?  That doesn't seem correct.

File: yarn_api_client/resource_manager.py [hunk 8]
```diff
@@ -13,38 +13,34 @@ class ResourceManager(BaseYarnAPI):
...
-                    if check_is_active_rm(alt_address, alt_port):
-                        address, port = alt_address, alt_port
+            for endpoint in service_endpoints:
+                if check_is_active_rm(endpoint, timeout):
+                    active_service_endpoint = endpoint
+                    break
 
-        super(ResourceManager, self).__init__(address, port, timeout, kerberos_enabled)
-
-    def get_active_host_port(self):
```
Comment (kevin-bates) [inline yarn_api_client/resource_manager.py]: Why is this removed?  This is used by clients to determine which RM is being used within an HA configuration.  It should be renamed to `get_active_endpoint` and return the endpoint which could be constructed using `self.service_uri.to_url(None)`.

File: yarn_api_client/base.py [hunk 9]
```diff
@@ -6,43 +6,62 @@
...
 
     def _validate_configuration(self):
-        if self.address is None:
-            raise ConfigurationError('API address is not set')
-        elif self.port is None:
-            raise ConfigurationError('API port is not set')
+        if not self.service_uri:
+            raise ConfigurationError('API endpoint is not set')
+
+    def _contruct_url(self, api_path):
```
Comment (kevin-bates) [inline yarn_api_client/base.py]: Move to `to_url` method per previous comment.

File: yarn_api_client/hadoop_conf.py [hunk 10]
```diff
@@ -41,19 +37,18 @@ def check_is_active_rm(rm_web_host, rm_web_port):
     else:
         if response.getheader('Refresh', None) is not None:
             return False
-        return True
+    return True
 
 
-def get_resource_manager_host_port():
+def get_resource_manager_host_port(timeout=30):
```
Comment (kevin-bates) [inline yarn_api_client/hadoop_conf.py]: Seems like this should be renamed to `get_resource_manager_endpoint`

File: yarn_api_client/hadoop_conf.py [hunk 11]
```diff
@@ -62,34 +57,20 @@ def get_resource_manager_host_port():
 def get_jobhistory_host_port():
```
Comment (kevin-bates) [inline yarn_api_client/hadoop_conf.py]: Seems like `get_jobhistory_endpoint()` is more fitting now.

File: yarn_api_client/hadoop_conf.py [hunk 12]
```diff
@@ -62,34 +57,20 @@ def get_resource_manager_host_port():
...
-    value = parse(config_path, prop_name)
-    if value is not None:
-        host, _, port = value.partition(':')
-        return host, port
-    else:
-        return None
+    return parse(config_path, prop_name)
 
 
 def get_nodemanager_host_port():
```
Comment (kevin-bates) [inline yarn_api_client/hadoop_conf.py]: Ditto - seems like replacing `host_port` with `endpoint` is the right thing to do.

File: yarn_api_client/hadoop_conf.py [hunk 13]
```diff
@@ -62,34 +57,20 @@ def get_resource_manager_host_port():
...
-    if value is not None:
-        host, _, port = value.partition(':')
-        return host, port
-    else:
-        return None
+    return parse(config_path, prop_name)
 
 
-def get_webproxy_host_port():
+def get_webproxy_host_port(timeout=30):
```
Comment (kevin-bates) [inline yarn_api_client/hadoop_conf.py]: `host_port` to `endpoint`.

File: yarn_api_client/resource_manager.py (hunk 8 above)
Comment (dimon222) [reply yarn_api_client/resource_manager.py]: It seems to be not useful, as it's possible to take out the base class members service_endooint_uri

File: yarn_api_client/base.py (hunk 7 above)
Comment (dimon222) [reply yarn_api_client/base.py]: If user passes domain without scheme, then all lands inside path. 
Example - pass service_endpoint `localhost`

File: yarn_api_client/resource_manager.py (hunk 8 above)
Comment (kevin-bates) [reply yarn_api_client/resource_manager.py]: This will be used by applications that pass in multiple endpoints but then need to know (for logging and troubleshooting purposes) which of those endpoints is the actual location for where things are happening.

File: yarn_api_client/base.py (hunk 7 above)
Comment (kevin-bates) [reply yarn_api_client/base.py]: Interesting about hostname and path.  Thanks for the explanation.

While I agree its a thin wrapper, it does provide some encapsulation so we can make changes in the future.  Thanks.

Comment (dimon222) [reply yarn_api_client/base.py]: I tried to incorporate best of two ideas, check latest commit.

File: yarn_api_client/base.py (hunk 9 above)
Comment (dimon222) [reply yarn_api_client/base.py]: Adjusted

File: yarn_api_client/hadoop_conf.py (hunk 10 above)
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: Adjusted

File: yarn_api_client/hadoop_conf.py (hunk 13 above)
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: Adjusted

File: yarn_api_client/resource_manager.py (hunk 8 above)
Comment (dimon222) [reply yarn_api_client/resource_manager.py]: Brought it back

File: yarn_api_client/hadoop_conf.py (hunk 12 above)
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: Adjusted

File: yarn_api_client/hadoop_conf.py (hunk 11 above)
Comment (dimon222) [reply yarn_api_client/hadoop_conf.py]: Adjusted

**Comment by dimon222** [comment]
@kevin-bates I applied suggested corrections. Let me know, so I can apply rebase with `fixup` to make it one consolidated PR

File: yarn_api_client/base.py [hunk 14]
```diff
@@ -23,20 +41,9 @@ def __init__(self, service_endpoint=None, timeout=None, auth=None, verify=True):
...
-            self.service_uri_scheme = self.service_uri.scheme or 'http'
-            self.service_uri_hostname = self.service_uri.hostname or self.service_uri.path
-
-            if self.service_uri.port:
-                self.service_uri_postfix = ':' + str(self.service_uri.port)
-            else:
-                self.service_uri_postfix = ''
+            self.service_uri = Uri(service_endpoint)
         else:
             self.service_uri = None
```
[the remark this answers is not in the document]
Comment (dimon222) [reply yarn_api_client/base.py]: Mainly due to existing test with indirect variable assignment in here:
https://github.com/toidi/hadoop-yarn-api-python-client/pull/34/files#diff-d06796e77090d59038f12fe9b1260baeL62
However, I do completely agree that it makes sense to validate right in constructor and rewrite test to use proper constructor case (since its meant to be used like this afterall!). Let me know what you think.

[the remark this answers is not in the document]
Comment (kevin-bates) [reply yarn_api_client/base.py]: Thanks for the link.  So I suppose the original intent was that clients could update the connection information after creating the instance.  That seems highly problematic to me - especially from a supportability standpoint.  However, because this is python, they'd still be able to do that, so I see the need to retain `_validate_configuration()` and, because this methodology may be in use (i.e., using the instance more like a singleton to hit different YARN clusters), I'm thinking it might be best to leave things as they are.

**Comment by dimon222** [comment]
@kevin-bates alright, thanks. I've rebased with `fixup` into one single commit with full changes. @lresende please have a final look and let me know if anything else I missed.

**Comment by dimon222** [comment]
Incorporated hotfix from #37

**Comment by dimon222** [comment]
Any plan on this?

**Comment by lresende** [comment]
I am good with any direction/feedback @kevin-bates gives towards merging this, otherwise, I have some hard deadlines this week and should be able to get back to this next week.

**Comment by kevin-bates** [comment]
Thanks for the ping @dimon222 (and @lresende).  I think my biggest concern is that this PR breaks compatibility with existing clients.  Enterprise Gateway 2.x caps yarn-api-client to < 0.4, so I think we're good there.  However, EG 1.x uses yarn-api-client >= 0.3.3, so users installing or upgrading will break w/o explicit pre-install or downgrade of yarn-api-client.  

Here's a possible plan:
1. Merge #39 and cut release 0.3.6.
2. Update EG 1.x to cap < 0.4.0 and cut minor releases 1.3.0 (and probably 2.0.0rc3 after merge of https://github.com/jupyter/enterprise_gateway/pull/699) . (Sorry, this item is not necessarily germane, but _is_ an instance of a broken client.)
3. Merge this PR and cut release 0.4.0.
4. Update EG 2.x to use new methods and add dependency for yarn-api-client >= 0.4.0 (Again, not necessarily germane)

Any comments, concerns?

cc: @dimon222, @IMAM9AIS, @lresende, @toidi, or anyone else watching this repo

**Comment by dimon222** [comment]
@kevin-bates sounds good, however the decision on https://github.com/toidi/hadoop-yarn-api-python-client/issues/8#issuecomment-497988074 would be useful incase we decide to introduce more breaking changes. Basically - to what extend we plan to support compliance with Hadoop, and do we need to backport/support old apis. Right now we literally have a mix of new and old apis (not good)

**Comment by toidi** [comment]
it's ok to introduce backward-incompatible changes
but if we do so - major version bump should be considered 1.x.x

usually, apps have pinned version of requirements. hadoop yarn client is not an exception here

or I got something wrong?

**Comment by kevin-bates** [comment]
Yeah, I agree with a major version bump.  Since we have yet to release a major (1.0) version at all, I wasn't sure if this should be that "thing", but perhaps it should.

I'm not familiar with the REST API and its version history.  @dimon222 - could you please elaborate on what parts of this repo are using different versions of the api?  I would hope that the `v1` in the URL represents the version of the REST API and that responses from a given version are stable.

If we wanted to support "legacy" versions, which given (glacial) speed at which organizations move - perhaps we should, then we should come up with a 'contract' for how many back-releases we'd support.  If only two, then perhaps a 'legacy' indicator in each constructor is appropriate (to the referenced comment) - or we actually use the version indicator of the URL (if its even applicable).

**Comment by kevin-bates** [comment]
Ok.  @toidi @dimon222 @lresende - How about this?  
1. We build `0.3.7` from master (to get the latest http-policy fix).  
2. Merge this PR and take a day or two to check things out (test with Enterprise Gateway, etc.)
3. Create 1.0.0 based on this PR

Applications that use the former api will need to pin < 1.0, if they haven't already (EG has pinned).

I'd be happy to work on these tomorrow (get 1 and 2 done, prepare for 3 - assuming 2 is successful).

**Comment by dimon222** [comment]
@kevin-bates Wow, surprised you wrote right in time I've looked at it :100: 
This PR been stuck for quite long time. I've just remembered and came back to do review of endpoint compliance for existing implementation. 

I've picked the latest possible Hadoop REST spec and next are my findings. I need to know opinion on how we want to keep compatibility, as we have couple of endpoints that still have "unstable/experimental" stage. That means endpoint spec can change anytime. If we want to polish it still, we could probably make beta version release and run our tests/polish the outstanding points i mentioned below.

ResourceManager:
1. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Applications_API - Applications API got several new arguments. I couldn't backtrack, I guess quite long ago (over 5 years).
2.  https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Specific_Container_for_an_Application_Attempt_API - implementation missing
3. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Nodes_API - spec doesn't match (changed at some point in the past?)
4. (ALPHA) https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Application_State_API - we have GET and independent "KILL" implementation, but not support for generic PUT with setting any other state. I might assume its kinda incomplete spec implementation
5. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Reservation_API_List - implementation missing
6. (ALPHA) https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Delegation_Tokens_API - implementation missing
7. (ALPHA) https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Reservation_API_Create - implementation missing
8. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Reservation_API_Submit - implementation missing
9. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Reservation_API_Update - implementation missing
10.  https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Reservation_API_Delete - implementation missing
11. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Application_Timeouts_API - implementation missing
12. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Application_Timeout_API - implementation missing
13. https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Application_Timeout_Update_API - implementation missing

NodeManager, HistoryServer, ApplicationMaster - TBD, i still have to review spec for them.

**Comment by kevin-bates** [comment]
@dimon222 - thanks for the detailed response - these are good points.

>... as we have couple of endpoints that still have "unstable/experimental" stage. That means endpoint spec can change anytime. 

Could you elaborate on which endpoints have "unstable/experimental" stage?

Regarding compatibility, I think we just need to be sure we follow the REST API docs and, right now, with the exception of the Timeline Service, there is just one version of the REST API - `v1`. I don't think we can get caught up in the underlying version of Hadoop/YARN, nor should we, since this is what a REST API is all about - acting as a thin veneer over more specific programmatic APIs.  As such, we may want to consider adding a _version indicator_ (e.g., _`v1`_, _`v2`_, etc.) as a parameter to the various constructors, with the implementation constructing the correct requests based on that value.

Nearly all of the links you provide are to requests that are not implemented.  I don't know the original history behind this project but I think we should treat additional requests on an as needed basis.  The project I work on (Enterprise Gateway), requires very few methods from this project, but should we discover others are required, we would certainly contribute their implementations.  If there are requests that you need, then by all means, let's be sure and get them in.  The issue with adding requests that aren't necessarily needed is that they increase the surface area for test and support and, in the case of the ALPHA requests, are subject to change.

Here's what I see as next steps...

1. Produce release `0.3.7` from master.
2. Merge this PR.
3. Produce release `1.0.0a` (alpha) as you suggest.  Existing applications should probably cap their requirements to `<1.0.0`.
4. See what additional APIs are necessary, fine-tune existing ones (e.g., look at adding REST version indicator, etc.) and produce a beta or release-candidate (depending on our confidence).
5. After some time on the beta/release candidate produce `1.0.0`

Comments, thoughts?

**Comment by dimon222** [comment]
@kevin-bates
I'm referring to Alpha marked ones.

I'm good with plan tho.

**Comment by kevin-bates** [comment]
Thanks for the quick response.  Cool - I was thinking it was the YARN ALPHAs you were referring to, but wanted to make sure it wasn't anything of ours.  

Great.  I'll produce `0.3.7` today.  If there are objections to that from others, speak up now (or prepare to build `0.3.8` :smile:)

**Comment by dimon222** [comment]
Closed in favor of #43
